arangod/Agency/RestAgencyHandler.cpp

```diff
@@ -150,7 +150,92 @@ RestStatus RestAgencyHandler::handleTransient() {
   return RestStatus::DONE;
 }
 
+
+/// Poll from agency starting with start.
+/// The result is of shape
+/// { accepted:bool, result: { firstIndex, commitIndex, logs:[{index, log}] } }
+/// if !accepted, lost leadership
+/// else respond
+RestStatus RestAgencyHandler::pollIndex(
+  index_t const& start, double const& timeout) {
+  return waitForFuture(
+    _agent->poll(start, timeout)
+    .thenValue([this, start](std::shared_ptr<VPackBuilder> rb) {
+      VPackSlice res = rb->slice();
+      if (res.hasKey("accepted")) {
+        TRI_ASSERT(res.hasKey("result"));
+        VPackBuffer<uint8_t> payload;
+        VPackBuilder builder(payload);
+        {
+          VPackObjectBuilder ob(&builder);
+          builder.add(StaticStrings::Error, VPackValue(false));
+          builder.add("code", VPackValue(int(ResponseCode::OK)));
+          builder.add(VPackValue("result"));
+          VPackObjectBuilder r(&builder);
+          VPackSlice slice = res.get("result");
+          auto const firstIndex = slice.get("firstIndex");
+          builder.add("commitIndex", slice.get("commitIndex"));
+          builder.add("firstIndex", firstIndex);
+          if (slice.hasKey("log")) {
+            VPackSlice logs = slice.get("log");
+            builder.add(VPackValue("log"));
+            VPackArrayBuilder a(&builder);
+            for (size_t i = start - firstIndex.getNumber<uint64_t>(); i < logs.length(); i++) {
+              builder.add(logs[i]);
+            }
+          } else {
+            builder.add("readDB", slice.get("readDB"));
+          }
+        }
+        resetResponse(rest::ResponseCode::ACCEPTED);
+        response()->setPayload(std::move(payload));
+      } else {
+        generateError(
+          rest::ResponseCode::SERVICE_UNAVAILABLE,
+          TRI_ERROR_HTTP_SERVER_ERROR, "No leader");
+      }
+    })
+    .thenError<VPackException>([this](VPackException const& e) {
+      generateError(Result{e.errorCode(), e.what()});
+    })
+    .thenError<std::exception>([this](std::exception const& e) {
+      generateError(
+        rest::ResponseCode::SERVER_ERROR, TRI_ERROR_HTTP_SERVER_ERROR, e.what());
+    }));
+
+}
+
+
+RestStatus RestAgencyHandler::handlePoll() {
+
+  // GET only
+  if (_request->requestType() != rest::RequestType::GET) {
+    return reportMethodNotAllowed();
+  }
+
+  // WARNING ////////////////////////////////////////////////////
+  // Leader only
+  if (!_agent->leading()) {  // Redirect to leader
+    if (_agent->leaderID() == NO_LEADER) {
```

Comment:
I vaguely remember a discussion we had as to whether we should allow followers to respond to this request. Did we not say that a follower could respond but should say it is a follower? I forgot the details.

Reply:
yes and this is unrelated. 503.